File: go_server/routes/uploads.go
package routes

import (
	"fmt"
	"net/http"
	"os"
	"path/filepath"
	"strings"

	"github.com/gin-gonic/gin"
	"github.com/google/uuid"

	"github.com/disintegration/imaging"
	"github.com/ingradient/ingradient/go_server/config"
)

func createThumbnail(originalPath, thumbnailPath string) error {
	// 이미지를 연다.
	img, err := imaging.Open(originalPath)
	if err != nil {
		return fmt.Errorf("failed to open image: %w", err)
	}

	// 예시: 150x150 크기로 썸네일 생성 (비율 유지가 필요하면 다른 함수 사용 가능)
	thumb := imaging.Thumbnail(img, 150, 150, imaging.Lanczos)

	// 썸네일 파일 저장
	err = imaging.Save(thumb, thumbnailPath)
	if err != nil {
		return fmt.Errorf("failed to save thumbnail: %w", err)
	}

	return nil
}

// UploadFile (POST /api/uploads/upload-file)
func UploadFile(c *gin.Context) {
	file, err := c.FormFile("file")
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	fileID := uuid.NewString()
	fname := fmt.Sprintf("%s_%s", fileID, file.Filename)

	folderPath := filepath.Join(config.UploadDir, "images")
	thumbnailFolder := filepath.Join(config.UploadDir, "thumbnails")
	os.MkdirAll(folderPath, 0755)
	os.MkdirAll(thumbnailFolder, 0755)

	fileLocation := filepath.Join(folderPath, fname)
	if err := c.SaveUploadedFile(file, fileLocation); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	thumbnailLocation := filepath.Join(thumbnailFolder, fname)

	// createThumbnail 호출
	if err := createThumbnail(fileLocation, thumbnailLocation); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Thumbnail generation failed: " + err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"id":                fileID,
		"filename":          file.Filename,
		"fileLocation":      fileLocation,
		"thumbnailLocation": thumbnailLocation,
	})
}

// UploadTempFile (POST /api/uploads/upload-temp)
func UploadTempFile(c *gin.Context) {
	file, err := c.FormFile("file")
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	sessionID := c.PostForm("session_id")
	if sessionID == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "session_id is required"})
		return
	}

	tmpSessionFolder := filepath.Join(config.TmpFolder, sessionID)
	os.MkdirAll(tmpSessionFolder, 0755)

	fileID := uuid.NewString()
	tmpFilename := fmt.Sprintf("%s_%s", fileID, file.Filename)
	tmpFilePath := filepath.Join(tmpSessionFolder, tmpFilename)

	if err := c.SaveUploadedFile(file, tmpFilePath); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"fileId":       fileID,
		"filename":     file.Filename,
		"tempLocation": tmpFilePath,
	})
}

// CommitUploads (POST /api/uploads/commit-uploads)
func CommitUploads(c *gin.Context) {
	// multipart/form-data로부터 session_id와 file_ids를 읽음.
	sessionID := c.PostForm("session_id")
	fileIDs := c.PostFormArray("file_ids")

	// 로그로 값 확인
	fmt.Printf("CommitUploads: session_id=%s, file_ids=%v\n", sessionID, fileIDs)

	// session_id 혹은 file_ids가 없으면 에러 응답
	if sessionID == "" || len(fileIDs) == 0 {
		c.JSON(http.StatusBadRequest, gin.H{"error": "session_id and file_ids are required"})
		return
	}

	finalFolder := filepath.Join(config.UploadDir, "images")
	thumbnailFolder := filepath.Join(config.UploadDir, "thumbnails")
	os.MkdirAll(finalFolder, 0755)
	os.MkdirAll(thumbnailFolder, 0755)

	tmpSessionFolder := filepath.Join(config.TmpFolder, sessionID)
	movedFilesInfo := []map[string]string{}

	entries, err := os.ReadDir(tmpSessionFolder)
	if err != nil {
		fmt.Printf("CommitUploads: error reading tmp folder %s: %v\n", tmpSessionFolder, err)
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	fmt.Printf("CommitUploads: found %d entries in tmp folder\n", len(entries))

	for _, fileID := range fileIDs {
		var matchedFileName string
		for _, e := range entries {
			// 파일 이름이 fileID로 시작하는지 확인
			if !e.IsDir() && len(e.Name()) > len(fileID) && e.Name()[:len(fileID)] == fileID {
				matchedFileName = e.Name()
				break
			}
		}
		if matchedFileName == "" {
			fmt.Printf("CommitUploads: No matching file for fileID: %s\n", fileID)
			continue
		}
		tmpFilePath := filepath.Join(tmpSessionFolder, matchedFileName)
		finalFilePath := filepath.Join(finalFolder, matchedFileName)

		// 파일 이동
		if err := os.Rename(tmpFilePath, finalFilePath); err != nil {
			fmt.Printf("CommitUploads: rename error %s -> %s: %v\n", tmpFilePath, finalFilePath, err)
		} else {
			fmt.Printf("CommitUploads: moved file %s to %s\n", tmpFilePath, finalFilePath)
		}

		// 썸네일 생성 (필요 시 createThumbnail 함수 호출)
		thumbnailPath := filepath.Join(thumbnailFolder, matchedFileName)
		// 만약 createThumbnail 같은 함수가 있다면 호출
		// err := createThumbnail(finalFilePath, thumbnailPath)
		// if err != nil { ... }

		// 파일 이름을 uuid_원본이름 형태에서 원본 이름만 추출
		tokens := strings.SplitN(matchedFileName, "_", 2)
		actualFilename := matchedFileName
		if len(tokens) == 2 {
			actualFilename = tokens[1]
		}

		movedFilesInfo = append(movedFilesInfo, map[string]string{
			"id":                fileID,
			"filename":          actualFilename,
			"fileLocation":      finalFilePath,
			"thumbnailLocation": thumbnailPath,
		})
	}

	fmt.Printf("CommitUploads: movedFilesInfo: %+v\n", movedFilesInfo)
	c.JSON(http.StatusOK, gin.H{
		"status":     "ok",
		"movedFiles": movedFilesInfo,
	})
}

// CancelUploads (DELETE /api/uploads/cancel-uploads)
func CancelUploads(c *gin.Context) {
	sessionID := c.PostForm("session_id")
	if sessionID == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "session_id is required"})
		return
	}
	tmpSessionFolder := filepath.Join(config.TmpFolder, sessionID)
	os.RemoveAll(tmpSessionFolder)
	c.JSON(http.StatusOK, gin.H{"status": "ok"})
}

// DownloadFile (GET /api/uploads/download/:filename)
func DownloadFile(c *gin.Context) {
	filename := c.Param("filename")
	filePath := filepath.Join(config.UploadDir, filename)
	if _, err := os.Stat(filePath); os.IsNotExist(err) {
		c.JSON(http.StatusNotFound, gin.H{"error": "File not found"})
		return
	}

	// 실제 파일 다운로드 응답
	c.File(filePath)
}

File: server/api/uploads.py
import asyncio
from fastapi import (
    APIRouter, BackgroundTasks, UploadFile, File, Form, Depends, Request,
    WebSocket, WebSocketDisconnect
)
import os
import shutil
import uuid
from typing import Optional, List
from server.utils.file_utils import create_thumbnail, delete_file
from server.core.config import UPLOAD_DIR, TMP_FOLDER
from server.utils.string_utils import to_camel_case

# ConnectionManager import (경로는 실제 프로젝트 구조에 맞게 수정하세요)
from server.core.websockets import manager 

router = APIRouter()

@router.post("/upload-file")
async def upload_file(file: UploadFile = File(...)):
    file_id = str(uuid.uuid4()).replace("-", "")
    fname = f"{file_id}_{file.filename}"

    folder_path = os.path.join(UPLOAD_DIR, "images")
    thumbnail_folder = os.path.join(UPLOAD_DIR, "thumbnails")

    os.makedirs(folder_path, exist_ok=True)
    os.makedirs(thumbnail_folder, exist_ok=True)
    
    file_location = os.path.join(folder_path, fname)
    thumbnail_location = os.path.join(thumbnail_folder, fname)

    with open(file_location, "wb") as f:
        shutil.copyfileobj(file.file, f)
    
    thumbnail_location = create_thumbnail(file_location, thumbnail_location)

    response = {
        "id": file_id,
        "filename": file.filename,
        "fileLocation": file_location,
        "thumbnailLocation": thumbnail_location,
    }
    return {to_camel_case(k): v for k, v in response.items()}

@router.post("/upload-temp")
async def upload_temp_file(file: UploadFile = File(...), session_id: str = Form(...)):
    tmp_session_folder = os.path.join(TMP_FOLDER, session_id)
    os.makedirs(tmp_session_folder, exist_ok=True)

    file_id = str(uuid.uuid4()).replace("-", "")
    tmp_file_path = os.path.join(tmp_session_folder, f"{file_id}_{file.filename}")

    with open(tmp_file_path, "wb") as f:
        shutil.copyfileobj(file.file, f)

    return {"fileId": file_id, "filename": file.filename, "tempLocation": tmp_file_path}

def process_files_in_background(session_id: str, file_ids: List[str], loop: asyncio.AbstractEventLoop):
    """실제 파일 처리 로직 (백그라운드에서 실행될 함수)"""
    final_folder = os.path.join(UPLOAD_DIR, "images")
    thumbnail_folder = os.path.join(UPLOAD_DIR, "thumbnails")
    tmp_session_folder = os.path.join(TMP_FOLDER, session_id)

    os.makedirs(final_folder, exist_ok=True)
    os.makedirs(thumbnail_folder, exist_ok=True)

    moved_files_info = []
    total_files = len(file_ids)
    processed_count = 0

    try:
        for i, file_id in enumerate(file_ids):
            # ... (파일 찾기 및 이동 로직) ...
            matched_files = [f for f in os.listdir(tmp_session_folder) if f.startswith(f"{file_id}_")]
            if not matched_files:
                continue
            tmp_filename = matched_files[0]
            tmp_file_path = os.path.join(tmp_session_folder, tmp_filename)
            final_file_path = os.path.join(final_folder, tmp_filename)
            shutil.move(tmp_file_path, final_file_path)
            thumbnail_path = create_thumbnail(final_file_path, os.path.join(thumbnail_folder, tmp_filename))
            moved_files_info.append({
                "id": file_id,
                "filename": tmp_filename.split("_", 1)[1],
                "fileLocation": final_file_path,
                "thumbnailLocation": thumbnail_path,
            })
            processed_count += 1
            # ... (파일 처리 완료) ...

            # N개마다 또는 일정 시간마다 진행률 전송
            if processed_count % 10 == 0 or processed_count == total_files:
                print(f"[{session_id}] Progress: {processed_count}/{total_files}")
                # 👇 전달받은 loop를 사용합니다.
                future = asyncio.run_coroutine_threadsafe(
                    manager.send_progress(session_id, processed_count, total_files), loop
                )
                future.result() # 작업이 메인 루프에서 실행되기를 기다림 (timeout 추가 고려)

        # 완료 메시지 전송
        # 👇 전달받은 loop를 사용합니다.
        future = asyncio.run_coroutine_threadsafe(
            manager.send_complete(session_id, moved_files_info), loop
        )
        future.result()

    except Exception as e:
        print(f"Error during background processing: {e}")
        # 👇 전달받은 loop를 사용합니다.
        future = asyncio.run_coroutine_threadsafe(
            manager.send_error(session_id, str(e)), loop
        )
        future.result()
    finally:
        # 백그라운드 스레드에서 루프를 닫거나 관리할 필요는 없습니다. (메인 루프이므로)
        pass # 특별히 할 일 없음


@router.post("/commit-uploads")
async def commit_uploads(
    request: Request,
    background_tasks: BackgroundTasks
):
    form_data = await request.form()
    session_id = form_data.get("session_id")
    file_ids = form_data.getlist("file_ids")

    print(f"Received commit request for session: {session_id}, files: {file_ids} ({len(file_ids)})")

    if not session_id or not file_ids:
        print("Error: session_id or file_ids missing in form data.")
        return {"status": "error", "message": "Missing session_id or file_ids"}

    # 👇 현재 실행 중인 이벤트 루프를 가져옵니다.
    loop = asyncio.get_running_loop() 

    # 👇 백그라운드 작업에 루프를 전달합니다.
    background_tasks.add_task(process_files_in_background, session_id, file_ids, loop) 
    
    return {"status": "processing_started"}


@router.delete("/cancel-uploads")
async def cancel_uploads(session_id: str = Form(...)):
    tmp_session_folder = os.path.join(TMP_FOLDER, session_id)
    if os.path.exists(tmp_session_folder):
        shutil.rmtree(tmp_session_folder)
    return {"status": "ok"}

@router.get("/download/{filename}")
def download_file(filename: str):
    file_path = os.path.join(UPLOAD_DIR, filename)
    if os.path.exists(file_path):
        return {"file_path": file_path}
    return {"error": "File not found"}

@router.websocket("/ws/commit-progress/{session_id}")
async def websocket_endpoint(websocket: WebSocket, session_id: str):
    await manager.connect(websocket, session_id)
    try:
        while True:
            # WebSocket 연결을 유지하고, 클라이언트로부터 메시지를 받을 수 있음
            # (여기서는 주로 서버 -> 클라이언트 통신이므로, 오래 대기)
            data = await websocket.receive_text()
            print(f"Received message from {session_id}: {data}") # (옵션) 클라이언트 메시지 처리
    except WebSocketDisconnect:
        manager.disconnect(session_id)
        print(f"WebSocket disconnected for session: {session_id}")
    except Exception as e:
        print(f"Error in websocket for {session_id}: {e}")
        manager.disconnect(session_id)